api/src/main/java/org/apache/iceberg/expressions/LiteralSet.java

```diff
@@ -0,0 +1,212 @@
+
+/**
+ * Represents a set of literal values in an IN or NOT_IN predicate
+ * @param <T> The Java type of the value, which can be wrapped by a {@link Literal}
+ */
+public class LiteralSet<T> implements Set<T>, Serializable {
+  private final Set<T> literals;
+
+  @SuppressWarnings("unchecked")
+  LiteralSet(Set<Literal<T>> lits) {
+    Preconditions.checkArgument(lits == null || lits.size() > 1,
+        "The input literal set must include more than 1 element.");
+    literals = ImmutableSet.<T>builder().addAll(
+        lits.stream().map(
+            lit -> {
+              if (lit instanceof Literals.StringLiteral) {
+                return (T) new CharSeqWrapper((CharSequence) lit.value());
+              } else {
+                return lit.value();
+              }
+            }
+    ).iterator()).build();
+  }
+
+  @Override
+  public String toString() {
+    Iterator<T> it = literals.iterator();
+    if (!it.hasNext()) {
+      return "{}";
+    }
+
+    StringBuilder sb = new StringBuilder();
+    sb.append('{');
+    while (true) {
+      sb.append(it.next());
+      if (!it.hasNext()) {
+        return sb.append('}').toString();
+      }
+      sb.append(',').append(' ');
+    }
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+    LiteralSet<T> that = (LiteralSet<T>) other;
+    return literals.equals(that.literals);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(literals);
+  }
+
+  @Override
+  public boolean contains(Object object) {
+    return literals.contains(object);
+  }
+
+  @Override
+  public int size() {
+    return literals.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return literals.isEmpty();
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return literals.iterator();
+  }
+
+  @Override
+  public Object[] toArray() {
+    return literals.toArray();
+  }
+
+  @Override
+  public <X> X[] toArray(X[] a) {
+    return literals.toArray(a);
+  }
+
+  @Override
+  public boolean containsAll(Collection<?> c) {
+    return literals.containsAll(c);
+  }
+
+  @Override
+  public boolean add(T t) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean remove(Object o) {
+    throw new UnsupportedOperationException();
+  }
+
+
+  @Override
+  public boolean addAll(Collection<? extends T> c) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean retainAll(Collection<?> c) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean removeAll(Collection<?> c) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void clear() {
+    throw new UnsupportedOperationException();
+  }
+
+  static class CharSeqWrapper implements CharSequence, Serializable {
```

Comment:
I think this should use the existing `CharSequenceWrapper`.

Reply:
You can add the `CharSequence` methods to the other wrapper.

Reply:
Thanks for the suggestion! 
Will remove `CharSeqWrapper` and reuse the `CharSequenceWrapper` instead.